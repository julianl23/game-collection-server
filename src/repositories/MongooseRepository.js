import mongoose from 'mongoose';

class MongooseRepository {
  constructor() {
    const schema = this.getSchema();

    this.Model = mongoose.model(this.schemaName, schema);
  }

  get schema() {
    throw new Error(
      'Not implemented. Please provide a schema in your class definition.'
    );
  }

  get collectionName() {
    throw new Error(
      'Not implemented. Please provide a collectionName in your class definition.'
    );
  }

  get schemaName() {
    throw new Error(
      'Not implemented. Please provide a schemaName in you class definition'
    );
  }

  getSchema() {
    return new mongoose.Schema(this.schema, {
      collection: this.collectionName,
      timestamps: true
    });
  }

  insertMany(opts = {}) {
    return this.Model.insertMany(opts);
  }

  find(opts = {}) {
    return this.Model.find(opts);
  }

  create(opts = {}) {
    return this.Model.create(opts);
  }

  findOne(opts = {}) {
    return this.Model.findOne(opts);
  }

  findOneAndRemove(opts = {}) {
    return this.Model.findOneAndRemove(opts);
  }

  update(opts = {}) {
    return this.Model.update(opts);
  }

  updateOne(opts = {}) {
    return this.Model.updateOne(opts);
  }
}

export default MongooseRepository;
