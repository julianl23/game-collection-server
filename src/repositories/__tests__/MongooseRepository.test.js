import MongooseRepository from '../MongooseRepository';

jest.mock('mongoose', () => {
  return {
    model: jest.fn(() => {
      return {
        find: val => val,
        create: val => val,
        findOne: val => val,
        findOneAndRemove: val => val
      };
    }),
    Schema: jest.fn()
  };
});

describe('repositories', () => {
  describe('MongooseRepository', () => {
    class ImplementedTest extends MongooseRepository {
      get schema() {
        return {};
      }

      get collectionName() {
        return 'test';
      }

      get schemaName() {
        return 'test';
      }
    }

    it('throws an error when a schema is not implemented', () => {
      function testInit() {
        new MongooseRepository();
      }

      expect(testInit).toThrowErrorMatchingSnapshot();
    });

    it('throws an error when trying to access an unimplemented collectionName getter', () => {
      class Test extends MongooseRepository {
        get schema() {
          return {};
        }

        get schemaName() {
          return 'test';
        }
      }

      function testInit() {
        new Test();
      }

      expect(testInit).toThrowErrorMatchingSnapshot();
    });

    it('throws an error if you have not specified a schemaName', () => {
      class Test extends MongooseRepository {
        get schema() {
          return {};
        }

        get collectionName() {
          return 'test';
        }
      }

      expect(() => new Test()).toThrowError(
        'Not implemented. Please provide a schemaName in you class definition'
      );
    });

    it('calls the correct find function', () => {
      const testInstance = new ImplementedTest();
      const expected = {};

      const value = testInstance.find();

      expect(value).toEqual(expected);
    });

    it('calls the correct create function', () => {
      const testInstance = new ImplementedTest();
      const expected = {};

      const value = testInstance.create();

      expect(value).toEqual(expected);
    });

    it('calls the correct findOne function', () => {
      const testInstance = new ImplementedTest();
      const expected = {};

      const value = testInstance.findOne();

      expect(value).toEqual(expected);
    });

    it('calls the correct findOneAndRemove function', () => {
      const testInstance = new ImplementedTest();
      const expected = {};

      const value = testInstance.findOneAndRemove();

      expect(value).toEqual(expected);
    });
  });
});
