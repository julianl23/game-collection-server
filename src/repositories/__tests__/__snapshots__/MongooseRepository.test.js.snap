// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repositories MongooseRepository throws an error when a schema is not implemented 1`] = `"Not implemented. Please provide a schema in your class definition."`;

exports[`repositories MongooseRepository throws an error when trying to access an unimplemented collectionName getter 1`] = `"Not implemented. Please provide a collectionName in your class definition."`;
